### PR TITLE
Implement builder pattern

### PR DESCRIPTION
**Pull request by creekorful**
Hello @borntyping this PR implements the builder pattern to instantiate the logger & rework the selective logging feature. (implementation idea by @piegamesde)

Do you want me to increment minor version (no breaking change introduce) on this PR?

Don't hesitate to correct me If there's a better way, or if you have better ideas.

### CODE REVIEW

**Review by borntyping — APPROVED** [review]

**Comment by borntyping** [comment]
Yes - this should bump the major or minor version.

File: src/lib.rs [hunk 1]
```diff
@@ -4,23 +4,124 @@
 use chrono::Local;
 #[cfg(feature = "colored")]
 use colored::*;
-use log::{Level, Log, Metadata, Record, SetLoggerError};
+use log::{Level, LevelFilter, Log, Metadata, Record, SetLoggerError};
+use std::collections::HashMap;
+
+pub struct SimpleLogger {
+    /// The default logging level
+    default_level: Level,
```
Comment (piegamesde) [inline src/lib.rs:12]: Shouldn't this be a `LevelFilter` as well, in order to be able to default-disable logging?

File: src/lib.rs [hunk 2]
```diff
@@ -4,23 +4,124 @@
...
-            && (self.whitelisted_targets.is_empty()
-                || self
-                    .whitelisted_targets
-                    .iter()
-                    .any(|t| t == metadata.target()))
+        metadata.level().to_level_filter()
+            <= self
+                .target_levels
+                .get(metadata.target())
+                .cloned()
```
Comment (piegamesde) [inline src/lib.rs:123]: As `LevelFilter` is `Copy` cloning should not be necessary.

File: src/lib.rs [hunk 3]
```diff
@@ -4,23 +4,124 @@
...
+        self
+    }
+
+    /// 'Init' the actual logger, instantiate it and configure it,
+    /// this method MUST be called in order for the logger to be effective.
+    pub fn init(self) -> Result<(), SetLoggerError> {
+        #[cfg(all(windows, feature = "colored"))]
+        set_up_color_terminal();
+
+        log::set_max_level(LevelFilter::Trace);
```
Comment (piegamesde) [inline src/lib.rs:104]: Could this be specialized to `log::set_max_level(max(self.default_level, self.target_levels.values().max()));` (pseudo-code)?

File: examples/init.rs [hunk 4]
```diff
@@ -1,5 +1,7 @@
+use simple_logger::SimpleLogger;
+
 fn main() {
-    simple_logger::init().unwrap();
+    SimpleLogger::new().init().unwrap();
```
Comment (piegamesde) [inline examples/init.rs:4]: This makes me wonder. Is the pull request backwards compatible?

File: examples/init_with_target_level.rs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+use log::{Level, LevelFilter};
+use simple_logger::SimpleLogger;
+
+fn main() {
+    SimpleLogger::new()
+        .with_level(Level::Info)
+        .with_target_level("init_with_target_level", LevelFilter::Off)
```
Comment (piegamesde) [inline examples/init_with_target_level.rs:7]: **Bikeshedding:** I don't think the naming is really intuitive. FWIW, the `env_logger` crate uses `		.filter_level(log::LevelFilter::Trace).filter_module("rumqtt", log::LevelFilter::Info)`. I'd perhaps go with `with_level` and `with_module_level`.

**Review by piegamesde — CHANGES_REQUESTED** [review]
Thank you a lot for implementing this. I only got a few minor remarks and some bike-shedding.

File: examples/init.rs (hunk 4 above)
Comment (creekorful) [reply examples/init.rs:4]: Yes it is. The previous function are still there, using the builder pattern under the hood.

File: src/lib.rs (hunk 1 above)
Comment (creekorful) [reply src/lib.rs:12]: I find it weird first of being able to disable logger, but you're right.

File: src/lib.rs (hunk 2 above)
Comment (creekorful) [reply src/lib.rs]: I have used .copied() is that what you were thinking?

**Comment by creekorful** [comment]
Here's latest version with fixes & version bump :)
Thanks for the comment & suggestions!

**Review by piegamesde — APPROVED** [review]

**Comment by creekorful** [comment]
@borntyping Should we mark as deprecated the 'old' instantiation methods or do you want to keep them in the futur?

**Comment by borntyping** [comment]
Happy to deprecate the old functions.

File: src/lib.rs (hunk 2 above)
Comment (piegamesde) [reply src/lib.rs]: I don't think I've ever used this, as I usually make the copy implicit with some reference magic. But this looks fine.

Comment (creekorful) [reply src/lib.rs]: It doesn't work implicitly this time, dunno why.

**Comment by piegamesde** [comment]
Thank you for the constructive discussion and for implementing this. I'll now go ahead and clean up that messy temporary logging setup in my projects …

**Comment by creekorful** [comment]
Thank you both for the great discussion & the code review, it was fun to implement the feature.
I'm happy to have this functionality in simple_logger, IMO it's the best easiest logging library available and I didn't wanted to move on something else.

Cheers!

**Pull request merged**